### PR TITLE
feat/migrate-graph-creation-to-torch

### PR DESCRIPTION
**Pull request by JPXKQX**
This PR improves the performance of creating a graph:

- Update the **KNNEdges** & **CutOffEdges** to rely on `torch-cluster`. This method also supports moving the tensors to the GPU.
- **EdgeAttributeBuilder**'s are now implemented as message passing graphs. This method supports moving the tensors to the GPU.

The performance improvement for generating a graph n320 (~500k nodes) -> o96 (~40k nodes) -> n320 is:

version | Device | CutOffEdges(cutoff_factor=0.6) | CutOffEdges attribute | KNNEdges(num_nearest_neighbors=3) | KNNEdges attributes
-- | -- | -- | -- | -- | --
v3.0.0 | CPU | 2.25 s | 0.55 s | 11.83 s | 1.21 s
torch-support | CPU | 0.45 s | 0.31 s | 0.30 s | 0.54 s
torch-support. | 1 GPU | 0.19 s | 0.14 s | 0.15 s | 0.00 s

The main difference is that the Haversine distance is not supported in the torch-cluster. To solve this problem, we transform the 2D coordinates into 3D coordinates (sphere) before calculating the Euclidean distance. The edge direction calculation has also been refactored to use torch.tensor's instead of np.array's & scipy (using Rodrigues' rotation formula).

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-core?pullRequest=56) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-core?pullRequest=56) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: JPXKQX<br/>:x: bluefoxr<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-core?pullRequest=56) it.</sub>

File: graphs/CHANGELOG.md
Comment (HCookie) [inline graphs/CHANGELOG.md:1]: The changelog no longer needs to be manually updated.

File: graphs/src/anemoi/graphs/edges/attributes.py [hunk 1]
```diff
@@ -117,120 +109,76 @@ class EdgeLength(BaseEdgeAttribute):
...
-    -------
-    get_node_name(source_name, target_name)
-        Return the name of the node to copy.
-
-    get_raw_values(graph, source_name, target_name)
-        Computes the edge attribute from the source or target node attribute.
-
-    """
+class BaseAttributeFromNodeBuilder(BooleanBaseEdgeAttributeBuilder, ABC):
+    """Base class for propagating an attribute from the nodes to the edges."""
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/edges/attributes.py]: Adding a check to this class that `node_idx` is set would be good.

**Comment by HCookie** [comment]
I presume this is a breaking change, and should likely be marked as such.

**Comment by JPXKQX** [comment]
This is a non-breaking change. Current functionality will be maintained if `torch-cluster` is not installed.

**Comment by JPXKQX** [comment]
> I presume this is a breaking change, and should likely be marked as such.


This is a non-breaking change. Current functionality will be maintained if torch-cluster is not installed.

**Comment by HCookie** [comment]
@JPXKQX, what about the moving of the builder classes?

**Comment by JPXKQX** [comment]
This should work. The default config files use anemoi.graphs.nodes.XXX and this is still supported (the _init_ method is updated with the new paths).

File: graphs/src/anemoi/graphs/edges/attributes.py [hunk 2]
```diff
@@ -9,104 +9,96 @@
...
-    Attributes
-    ----------
-    norm : Optional[str]
-        Normalisation method. Options: None, "l1", "l2", "unit-max", "unit-range", "unit-std".
-    luse_rotated_features : bool
-        Whether to use rotated features.
+class BasePositionalBuilder(BaseEdgeAttributeBuilder, ABC):
+    node_attr_name: str = "x"
+    _idx_lat: int = 0
+    _idx_lon: int = 1
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/edges/attributes.py:69]: Why does this subclass ABC, yet provide no abstract methods?

File: graphs/src/anemoi/graphs/edges/attributes.py [hunk 3]
```diff
@@ -9,104 +9,96 @@
...
-            source_name in graph.node_types
-        ), f"Node \"{source_name}\" not found in graph. Optional nodes are {', '.join(graph.node_types)}."
-        assert (
-            target_name in graph.node_types
-        ), f"Node \"{target_name}\" not found in graph. Optional nodes are {', '.join(graph.node_types)}."
+        Returns
+        -------
+        torch.Tensor
+            Edge features.
+        """
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/edges/attributes.py]: Should this be an abstract method?

File: graphs/src/anemoi/graphs/edges/builders/base.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        attrs_config : DotDict
+            The configuration of the edge attributes.
+
+        Returns
+        -------
+        HeteroData
+            The graph with the edges.
+        """
+        if torch.cuda.is_available():
+            graph = graph.to("cuda")
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/edges/builders/base.py]: Are other backend's supported by torch_geometric?

File: graphs/src/anemoi/graphs/edges/builders/cutoff.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            "Using CutOff-Edges (with radius = %.1f km) between %s and %s.",
+            self.radius * EARTH_RADIUS,
+            self.source_name,
+            self.target_name,
+        )
+
+        try:
+            edge_index = self._compute_edge_index_pyg(source_nodes, target_nodes)
+        except ImportError:
+            import warnings
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/edges/builders/cutoff.py]: A proper check if importing torch_cluster would be better here, just so that a warning is not raised incorrectly.
https://docs.python.org/3/library/importlib.html#importlib.util.find_spec

File: graphs/src/anemoi/graphs/edges/builders/multi_scale.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        Update the graph with the edges.
+    """
+
+    VALID_NODES = [
+        TriNodes,
+        HexNodes,
+        LimitedAreaTriNodes,
+        LimitedAreaHexNodes,
+        StretchedTriNodes,
+    ]
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/edges/builders/multi_scale.py]: This should be a proper global variable outside of the class.

File: graphs/src/anemoi/graphs/edges/builders/multi_scale.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+    def compute_edge_index(self, source_nodes: NodeStorage, target_nodes: NodeStorage) -> torch.Tensor:
+        if source_nodes.node_type in [TriNodes.__name__, LimitedAreaTriNodes.__name__]:
+            source_nodes = self.add_edges_from_tri_nodes(source_nodes)
+        elif source_nodes.node_type in [HexNodes.__name__, LimitedAreaHexNodes.__name__]:
+            source_nodes = self.add_edges_from_hex_nodes(source_nodes)
+        elif source_nodes.node_type == StretchedTriNodes.__name__:
+            source_nodes = self.add_edges_from_stretched_tri_nodes(source_nodes)
+        else:
+            raise ValueError(f"Invalid node type {source_nodes.node_type}")
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/edges/builders/multi_scale.py]: This feels messy, could it be factored into a loop, or lookup?

File: graphs/src/anemoi/graphs/edges/attributes.py (hunk 2 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/edges/attributes.py:69]: It now inherits the abstract method from BaseEdgeAttributeBuilder.

File: graphs/src/anemoi/graphs/edges/builders/base.py (hunk 4 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/edges/builders/base.py]: No idea

File: graphs/src/anemoi/graphs/edges/builders/multi_scale.py (hunk 6 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/edges/builders/multi_scale.py]: Removed

File: graphs/src/anemoi/graphs/edges/builders/multi_scale.py (hunk 7 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/edges/builders/multi_scale.py]: Refactored!

**Comment by JPXKQX** [comment]
@HCookie This PR should be ready to merge. More unit tests. Global, Lam & stretched experiments were tested in MLflow. All comments addressed.

Let me know if you are missing anything.

File: graphs/src/anemoi/graphs/edges/attributes.py (hunk 2 above)
Comment (HCookie) [reply graphs/src/anemoi/graphs/edges/attributes.py:69]: As BaseEdgeAttributeBuilder is a subclass of ABC, does this need to be?

File: graphs/src/anemoi/graphs/edges/attributes.py [hunk 8]
```diff
@@ -117,120 +97,77 @@ class EdgeLength(BaseEdgeAttribute):
...
+    References
+    ----------
+    - https://www.movable-type.co.uk/scripts/latlong.html
     """
 
-    def __init__(self, norm: str | None = None, invert: bool = False) -> None:
-        super().__init__(norm)
-        self.invert = invert
+    def compute(self, x_i: torch.Tensor, x_j: torch.Tensor) -> torch.Tensor:
+        # Forward bearing. x_i, x_j must bez radians.
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/edges/attributes.py]:
```suggestion
        # Forward bearing. x_i, x_j must be radians.
```

File: graphs/src/anemoi/graphs/edges/attributes.py [hunk 9]
```diff
@@ -117,120 +97,77 @@ class EdgeLength(BaseEdgeAttribute):
...
+class BaseEdgeAttributeFromNodeBuilder(BaseBooleanEdgeAttributeBuilder, ABC):
+    """Base class for propagating an attribute from the nodes to the edges."""
 
+    def __init__(self, node_attr_name: str) -> None:
+        self.node_attr_name = node_attr_name
+        super().__init__()
+        if not hasattr(self, "node_idx"):
+            raise AttributeError(
+                "Classes inheriting from BaseEdgeAttributeFromNodeBuilder must set 'node_idx' attribute"
+            )
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/edges/attributes.py]: You've done a similar thing above with `BaseEdgeAttributeBuilder`, and provided it as None by default, why the difference?

File: graphs/src/anemoi/graphs/edges/attributes.py [hunk 10]
```diff
@@ -117,120 +97,77 @@ class EdgeLength(BaseEdgeAttribute):
...
-        if self.invert:
-            values = 1 - values
 
-        return values
+class BaseEdgeAttributeFromNodeBuilder(BaseBooleanEdgeAttributeBuilder, ABC):
+    """Base class for propagating an attribute from the nodes to the edges."""
 
+    def __init__(self, node_attr_name: str) -> None:
+        self.node_attr_name = node_attr_name
+        super().__init__()
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/edges/attributes.py:137]: super should come first

File: graphs/src/anemoi/graphs/nodes/attributes/boolean_op.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+from anemoi.graphs.nodes.attributes.base_attributes import BooleanBaseNodeAttribute
+
+LOGGER = logging.getLogger(__name__)
+MaskAttributeType = Union[str, Type["BooleanBaseNodeAttribute"]]
+
+
+class BooleanOperation(BooleanBaseNodeAttribute, ABC):
+    """Base class for boolean operations."""
+
+    def __init__(self, masks: MaskAttributeType | list[MaskAttributeType] = None) -> None:
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/nodes/attributes/boolean_op.py]: Why does this `masks` have a default value when it is not a valid value?

File: graphs/src/anemoi/graphs/nodes/attributes/masks.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+import torch
+from torch_geometric.data.storage import NodeStorage
+
+from anemoi.datasets import open_dataset
+from anemoi.graphs.nodes.attributes.base_attributes import BooleanBaseNodeAttribute
+
+LOGGER = logging.getLogger(__name__)
+
+
+class NonmissingZarrVariable(BooleanBaseNodeAttribute):
```
Comment (HCookie) [inline graphs/src/anemoi/graphs/nodes/attributes/masks.py]: Should this reference zarr?

File: graphs/src/anemoi/graphs/edges/attributes.py (hunk 2 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/edges/attributes.py:69]: I think it is not needed, but I always try to add it in case the class is abstract so it is explicitly defined.

File: graphs/src/anemoi/graphs/nodes/attributes/boolean_op.py (hunk 11 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/nodes/attributes/boolean_op.py]: No idea, it is removed now!

File: graphs/src/anemoi/graphs/nodes/attributes/masks.py (hunk 12 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/nodes/attributes/masks.py]: No

File: graphs/src/anemoi/graphs/edges/attributes.py (hunk 10 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/edges/attributes.py:137]: Done

File: graphs/src/anemoi/graphs/edges/attributes.py (hunk 8 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/edges/attributes.py]: Done!

File: graphs/src/anemoi/graphs/edges/attributes.py (hunk 9 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/edges/attributes.py]: I am not sure what you mean here

File: graphs/src/anemoi/graphs/edges/attributes.py (hunk 10 above)
Comment (JPXKQX) [reply graphs/src/anemoi/graphs/edges/attributes.py:137]: I had to come back to this as the `super().__init___()` check the node_attr_name is not None.

File: graphs/src/anemoi/graphs/edges/attributes.py (hunk 9 above)
Comment (HCookie) [reply graphs/src/anemoi/graphs/edges/attributes.py]: In the `BaseEdgeAttributeBuilder` you provided a None as a default value, and then checked to make sure it had been set (see https://github.com/ecmwf/anemoi-core/blob/e6c44717bf6944e7750e2fab1d2a65ec56ba062e/graphs/src/anemoi/graphs/edges/attributes.py#L33), here you check for an attribute. It seems as if you've gone for two ways to do the same thing.

Comment (JPXKQX) [reply graphs/src/anemoi/graphs/edges/attributes.py]: Ok, I understand now. It should be fixed now

**Review by HCookie — APPROVED** [review]
